File: app/src/main/java/com/example/companioniiit/add_reminder_page.java
package com.example.companioniiit;

import android.app.AlarmManager;
import android.app.DatePickerDialog;
import android.app.PendingIntent;
import android.app.TimePickerDialog;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.TimePicker;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

import java.util.Calendar;

public class add_reminder_page extends AppCompatActivity {

    private EditText titleEditText, descriptionEditText;
    private Button datePickerButton, timePickerButton, saveReminderButton;
    private int year, month, day, hour, minute;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_reminder_page);

        titleEditText = findViewById(R.id.titleEditText);
        descriptionEditText = findViewById(R.id.descriptionEditText);
        datePickerButton = findViewById(R.id.datePickerButton);
        timePickerButton = findViewById(R.id.timePickerButton);
        saveReminderButton = findViewById(R.id.saveReminderButton);

        datePickerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Calendar calendar = Calendar.getInstance();
                year = calendar.get(Calendar.YEAR);
                month = calendar.get(Calendar.MONTH);
                day = calendar.get(Calendar.DAY_OF_MONTH);

                DatePickerDialog datePickerDialog = new DatePickerDialog(add_reminder_page.this,
                        new DatePickerDialog.OnDateSetListener() {
                            @Override
                            public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                                add_reminder_page.this.year = year;
                                add_reminder_page.this.month = monthOfYear;
                                add_reminder_page.this.day = dayOfMonth;
                            }
                        }, year, month, day);
                datePickerDialog.show();
            }
        });

        timePickerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Calendar calendar = Calendar.getInstance();
                hour = calendar.get(Calendar.HOUR_OF_DAY);
                minute = calendar.get(Calendar.MINUTE);

                TimePickerDialog timePickerDialog = new TimePickerDialog(add_reminder_page.this,
                        new TimePickerDialog.OnTimeSetListener() {
                            @Override
                            public void onTimeSet(TimePicker view, int hourOfDay, int minute) {
                                add_reminder_page.this.hour = hourOfDay;
                                add_reminder_page.this.minute = minute;
                            }
                        }, hour, minute, true);
                timePickerDialog.show();
            }
        });

        saveReminderButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String title = titleEditText.getText().toString();
                String description = descriptionEditText.getText().toString();

                if (title.isEmpty() || description.isEmpty()) {
                    Toast.makeText(add_reminder_page.this, "Title and Description cannot be empty", Toast.LENGTH_SHORT).show();
                    return;
                }

                Calendar calendar = Calendar.getInstance();
                calendar.set(year, month, day, hour, minute);

                saveReminderToFirebase(title, description, calendar.getTimeInMillis());
                scheduleNotification(title, description, calendar.getTimeInMillis());
            }
        });
    }

    private void saveReminderToFirebase(String title, String description, long time) {
        DatabaseReference databaseReference = FirebaseDatabase.getInstance().getReference("reminders");
        String reminderId = databaseReference.push().getKey();

        Reminder reminder = new Reminder(reminderId, title, description, time);
        databaseReference.child(reminderId).setValue(reminder).addOnCompleteListener(new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    Toast.makeText(add_reminder_page.this, "Reminder saved", Toast.LENGTH_SHORT).show();
                    finish();
                } else {
                    Toast.makeText(add_reminder_page.this, "Failed to save reminder", Toast.LENGTH_SHORT).show();
                }
            }
        });
    }

    private void scheduleNotification(String title, String description, long time) {
        Intent intent = new Intent(this, ReminderBroadcastReceiver.class);
        intent.putExtra("title", title);
        intent.putExtra("description", description);

        PendingIntent pendingIntent = PendingIntent.getBroadcast(this, 0, intent, PendingIntent.FLAG_UPDATE_CURRENT);

        AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
        if (alarmManager != null) {
            alarmManager.setExact(AlarmManager.RTC_WAKEUP, time, pendingIntent);
        }
    }
}

File: app/src/main/java/com/example/companioniiit/AddreminderPage.kt
package com.example.companioniiit

import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.util.Log
import android.widget.ArrayAdapter
import android.widget.TimePicker
import android.widget.Toast
import androidx.activity.enableEdgeToEdge
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.NotificationManagerCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.workDataOf
import com.example.companioniiit.databinding.ActivityAddreminderPageBinding
import com.example.companioniiit.databinding.ReminderDialogBinding
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.text.SimpleDateFormat
import java.time.Year
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

class AddreminderPage : AppCompatActivity() {

    private  lateinit var binding: ActivityAddreminderPageBinding
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
         binding = ActivityAddreminderPageBinding.inflate(layoutInflater)
        setContentView(binding.root)

        binding.Addreminder.setOnClickListener {
            addReminder()
        }

        }

    private fun addReminder() {
        if(Build.VERSION.SDK_INT >= 33 && !NotificationManagerCompat.from(this).areNotificationsEnabled()){
            showNotificationPermissionDialog()
        } else{
            addReminderDialog()
        }

    }

    private fun addReminderDialog() {
        val dialogBinding= ReminderDialogBinding.bind(layoutInflater.inflate(R.layout.reminder_dialog,null))
        val dialog=AlertDialog.Builder(this)
            .setView(dialogBinding.root)
            .show()

        dialogBinding.ReminderType.adapter = ArrayAdapter(this,android.R.layout.simple_list_item_1,resources.getStringArray(R.array.reminderTypes))
        dialogBinding.cancelbtn.setOnClickListener {
            dialog.dismiss()
        }


        val pickedDate= Calendar.getInstance()
        dialogBinding.selectbtn.setOnClickListener {

            val year=pickedDate.get(Calendar.YEAR)
            val month=pickedDate.get(Calendar.MONTH)
            val day=pickedDate.get(Calendar.DATE)
            val hour=pickedDate.get(Calendar.HOUR_OF_DAY)
            val minute=pickedDate.get(Calendar.MINUTE)

            DatePickerDialog(this,
                DatePickerDialog.OnDateSetListener{view, year, month, dayOfMonth ->
                        // show timepicker
                        TimePickerDialog(this,
                            TimePickerDialog.OnTimeSetListener{view, hourOfDay, minute ->
                                pickedDate.set(year,month,dayOfMonth,hourOfDay, minute)
                                Log.d("DateAndTime","Picked Date and time $pickedDate")
                                dialogBinding.pickedDateAndTime.text=getCurrentDateAndTime(pickedDate.timeInMillis)
                            },
                            hour,
                            minute,
                            false
                        ).show()


                }
                ,year,month,day)
                .show()


        }

        dialogBinding.submitbtn.setOnClickListener{
            if (dialogBinding.edittitle.text.isNullOrEmpty()){
                dialogBinding.edittitle.requestFocus()
                dialogBinding.edittitle.setError("Please provide Title or description")

            }else if (dialogBinding.pickedDateAndTime.text== resources.getString(R.string.date_and_time)){
                dialogBinding.pickedDateAndTime.setError("please select date and time!!!")


            }
            else{
                // adding reminder
                val timeDelayInSeconds = (pickedDate.timeInMillis / 1000L) - (Calendar.getInstance().timeInMillis / 1000L)
                if(timeDelayInSeconds<0){
                    Toast.makeText(this,"Can not set reminders for past dates",Toast.LENGTH_LONG).show()
                    return@setOnClickListener
                }
                // work manager to set reminder
                createWorkRequest(dialogBinding.edittitle.text.toString(),resources.getStringArray(R.array.reminderTypes)[dialogBinding.ReminderType.selectedItemPosition],
                    timeDelayInSeconds)
                Toast.makeText(this,"Reminder Added",Toast.LENGTH_LONG).show()
                dialog.dismiss()

            }
        }

    }

    private fun createWorkRequest(title: String, reminderType: String,     delay: Long) {
        val reminderWorkRequest = OneTimeWorkRequestBuilder<ReminderWorker>()
            .setInitialDelay(delay, TimeUnit.SECONDS)
            .setInputData(
                workDataOf(
                    "Title" to "Todo: $reminderType",
                    "Message" to title
                )
            )
            .build()
        WorkManager.getInstance(this).enqueue(reminderWorkRequest)
    }

    private fun getCurrentDateAndTime(millis:Long):String{
        return SimpleDateFormat("dd-MM-yyyy hh:mm a", Locale.getDefault()).format(Date(millis))




    }


    private fun showNotificationPermissionDialog() {
       MaterialAlertDialogBuilder(this,
       com.google.android.material.R.style.MaterialAlertDialog_Material3
       )
           .setTitle("Notification Permission")
           .setMessage("We need permission to send notifications")
           .setPositiveButton("Ok"){ dialog, which ->
               // notification launcher
               if(Build.VERSION.SDK_INT >= 33){
                   notificationLauncher.launch(android.Manifest.permission.POST_NOTIFICATIONS)
               }
           }
           .setNegativeButton("Cancel",null)
           .show()
    }

    val notificationLauncher = registerForActivityResult(ActivityResultContracts.RequestPermission()){
        isGranted ->
        if(isGranted){
            if(Build.VERSION.SDK_INT >= 33){
                if(shouldShowRequestPermissionRationale(android.Manifest.permission.POST_NOTIFICATIONS)){
                    showNotificationPermissionDialog()
                }else{
                    showSettingDialog()
                }



            }

        }else{
            Toast.makeText(this,"Notification permission granted", Toast.LENGTH_SHORT).show()
        }


    }

    private fun showSettingDialog() {
       MaterialAlertDialogBuilder(this,

           com.google.android.material.R.style.MaterialAlertDialog_Material3)
           .setTitle("Notification Permission")
           .setMessage("We need permission to send notifications")
           .setPositiveButton("Ok"){ _ , _ ->
               val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
               intent.data = Uri.parse("package:${this.applicationContext.packageName}")
               startActivity(intent)
           }
           .setNegativeButton("Cancel",null)
           .show()

           }

    }



